src/kademlia/dht_storage.cpp

```diff
@@ -532,6 +556,36 @@ namespace
 			if (!peers.empty() && peers.capacity() / peers.size() >= 4u)
 				peers.shrink_to_fit();
 		}
+
+		void refresh_infohashes_sample()
+		{
+			time_point const now = aux::time_now();
+			int const interval = clamp(m_settings.sample_infohashes_interval, 0, 21600);
+
+			std::size_t const max_count
+				= std::size_t(clamp(m_settings.max_infohashes_sample_count, 0, 20));
+			std::size_t const count = std::min(max_count, m_map.size());
+
+			if (interval > 0
+				&& m_infohashes_sample.created + seconds(interval) > now
+				&& m_infohashes_sample.samples.size() >= max_count)
+				return;
+
+			std::vector<node_id> keys;
+			for (auto const& t : m_map)
+				keys.push_back(t.first);
```

Comment:
Rather than copy all in info hashes into a vector and shuffle that, it would be more efficient to pick hashes with the appropriate probability like get_peers does.